File: 278.first-bad-version.py
#
# @lc app=leetcode id=278 lang=python3
#
# [278] First Bad Version
#

# @lc code=start
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        lp, rp = 1, n
        while lp <= rp:
            mid = lp + (rp-lp)//2
            if isBadVersion(mid) == True:
                rp = mid - 1
            else:
                lp = mid + 1
        return lp
# @lc code=end

File: 278.first-bad-version.java
/*
 * @lc app=leetcode id=278 lang=java
 *
 * [278] First Bad Version
 */

// @lc code=start
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int lp = 0;
        int rp = n;
        int rtn = -1;

        while(lp <= rp) {
            n = lp + (rp - lp) / 2;
            if (isBadVersion(n)) {
                rtn = n;
                rp = n - 1;
            } else {
                lp = n + 1;
            }
        }
        return rtn;
    }
}
// @lc code=end

